python-package/lightgbm/sklearn.py

```diff
@@ -1187,16 +1193,16 @@ def fit(
         init_score=None,
         group=None,
         eval_set=None,
-        eval_names=None,
+        eval_names: Optional[List[str]] = None,
         eval_sample_weight=None,
         eval_init_score=None,
         eval_group=None,
-        eval_metric=None,
-        eval_at=(1, 2, 3, 4, 5),
+        eval_metric: Optional[_LGBM_ScikitEvalMetricType] = None,
+        eval_at: Iterable[int] = (1, 2, 3, 4, 5),
```

Comment:
I think this should be `Sequence[int]`, given that for example `range(5)` is an iterable but is not a valid type for this.

Reply:
🤩 Excellent point. I was just following the docstring and didn't consider this. Thanks very much for noting it!!!

I tried the following from the root of the repo, just to see what would happen:

<details><summary>sample code (click me)</summmary>

```python
from pathlib import Path
import numpy as np
import lightgbm as lgb
from sklearn.datasets import load_svmlight_file

rank_example_dir = Path('examples/lambdarank')
X_train, y_train = load_svmlight_file(str(rank_example_dir / 'rank.train'))
X_test, y_test = load_svmlight_file(str(rank_example_dir / 'rank.test'))
q_train = np.loadtxt(str(rank_example_dir / 'rank.train.query'))
q_test = np.loadtxt(str(rank_example_dir / 'rank.test.query'))
gbm = lgb.LGBMRanker(n_estimators=10)
gbm.fit(
    X_train,
    y_train,
    group=q_train,
    eval_set=[(X_test, y_test)],
    eval_group=[q_test],
    eval_at=range(3),
    callbacks=[
        lgb.early_stopping(10),
        lgb.reset_parameter(learning_rate=lambda x: max(0.01, 0.1 - 0.01 * x))
    ]
)
```

</details>

And you're right...passing a range for `eval_at` causes a failure when serializing the parameters to string to pass them through the C API functions.

```text
  File ".../site-packages/lightgbm/basic.py", line 326, in param_dict_to_str
    raise TypeError(f'Unknown type of parameter:{key}, got:{type(val).__name__}')
TypeError: Unknown type of parameter:eval_at, got:range
```

Given that, I think the hint here should be even stricter than `typing.Sequence`. Since this keyword argument is passed directly through to `params` and there's no other code in LightGBM manipulating its value, I think it can only accept values that are valid for `lightgbm.basic.param_dict_to_str()`.

For `eval_at`, I think that means only a list of ints or tuple of ints is valid. `param_dict_to_str()` supports `list`, `tuple`, and `set`, but `set` isn't appropriate for `eval_at` because sets aren't iterable (e.g. don't have any ordering).

https://github.com/microsoft/LightGBM/blob/3d4e08e1d8be9763f5d55180e3faf1e11eb20787/python-package/lightgbm/basic.py#L323

I just pushed https://github.com/microsoft/LightGBM/pull/5460/commits/81c234f27f35478ade88857d888190b8554ef47d which:

* sets the hint for `eval_at` to `Union[List[int], Tuple[int]]`
* replaces use of the word "iterable" in the relevant docstrings with "list or tuple of int"

Reply:
@jmoralez I won't merge this until you have a chance to respond, since what I did here is slightly different than what you suggested.